src/core/storage/sframe_interface/unity_sframe.cpp

```diff
@@ -1200,6 +1202,80 @@ unity_sframe::random_split(float percent, uint64_t random_seed, bool exact) {
   return logical_filter_split(logical_filter_array);
 }
 
+std::shared_ptr<unity_sframe_base> unity_sframe::shuffle() {
+  log_func_entry();
+
+  std::vector<std::string> column_names = this->column_names();
+  const size_t num_buckets = (this->size() / SFRAME_SHUFFLE_BUCKET_SIZE) + 1;
+
+  // Create a column of random ints between 0 and (num_buckets - 1).
+  auto temp_groupby_column = std::static_pointer_cast<unity_sarray>(
+    unity_sarray::make_uniform_int_array(this->size(), num_buckets));
+  const std::string rand_int_column_name("Random Ints");
+  std::shared_ptr<unity_sframe> temp(new unity_sframe());
+  temp->add_column(temp_groupby_column, rand_int_column_name);
+
+  // Pack columns so we can group by concatenate
+  std::shared_ptr<unity_sarray_base> packed_columns = this->pack_columns(column_names,
+                                                                         column_names,
+                                                                         flex_type_enum::LIST,
+                                                                         turi::flex_undefined());
+  const std::string packed_data_column_name("Packed Data");
+  temp->add_column(packed_columns, packed_data_column_name);
+
+  // Group by concatenate on the random int column. This randomly bucketizes.
+  const std::string buckets_column_name("Buckets");
+  std::shared_ptr<unity_sframe_base> bucketized_sframe = temp->groupby_aggregate({rand_int_column_name},
+                                                                                 {{packed_data_column_name}},
+                                                                                 {buckets_column_name},
+                                                                                 {"__builtin__concat__list__"});
+  std::shared_ptr<unity_sarray_base> bucketized_sarray = bucketized_sframe->select_column(buckets_column_name);
+
+  // Shuffle each bucket
+  size_t num_threads = thread::cpu_count();
+  gl_sarray_writer writer(flex_type_enum::LIST, num_threads);
+  gl_sarray gl_bucketized_sarray = gl_sarray(bucketized_sarray);
+
+  in_parallel([&](size_t thread_id, size_t n_threads) {
+      size_t idx_start = (gl_bucketized_sarray.size() * thread_id) / n_threads;
+      size_t idx_end = (gl_bucketized_sarray.size() * (thread_id + 1) ) / n_threads;
+
+      gl_sarray_range ra = gl_bucketized_sarray.range_iterator(idx_start, idx_end);
+      auto cur_bucket = ra.begin();
+      unsigned int seed = std::chrono::system_clock::now().time_since_epoch().count();
```

Comment:
why don't we use the pure_random function? Just curious.

Reply:
I agree exposing a seed parameter to the end user would be nice. However it's not possible. We have to use groupby in order to get a linear time solution. Groupby is non-deterministic. So it doesn't take a seed parameter. So we can't either.

Reply:
I think groupby is deterministic if the hash key is generated as the same.

Reply:
okay, I see, elements in the bucket are not. Well, I agree. It's hard.

Reply:
The only way I can think of is to do the indirect sort on each bucket (sort their index). And then shuffle the indirectly sorted index vector.

Reply:
I don't know what you mean by "indirect sort on each bucket". However if you're doing any sorting, then I don't think the solution can have a linear running time.

Reply:
for each hash key, you have a bucket, which is your list (vector<flex>) here. When you perform indirect sort, you have a separate `vector<size_t> indices`, which represents the index in the bucket.

```
indices.sort(ii, jj) { bucket[ii] < bucket[jj]; }
```

This will save extra copies on elements in bucket.

Nlog(N/B) time is not bad. You already save tons of random reads that involve io.

Reply:
The worst case is NlogN is when all elements go into the same bucket. It depends on whether it's worthy of sacrificing extra NlogN - N **memory read** for a reproducible model. 

For most of our inputs, N is less than 10k, log(10k) is 9. Assume each memory random read is 1us, and io read is 1ms.

Assume a sframe can be fully fit into memory, there's no io read from disk. The read will be N * 1us = 100ms.

The worst penalty for memory read is 80k more reads if we sort the bucket. you pay extra 800 ms. Not that bad.

If sframe is very large, more than 2G, which is the cache manager capacity. we may have some read that needs to be load from disk. Let's say only 10% of them will miss the cache and force the load from disk, you pay N*10%*1ms. If N is 10k, then it's about 1000ms.